src/ripple/app/misc/AMMHelpers.h

```diff
@@ -147,12 +149,223 @@ withinRelativeDistance(Amt const& calc, Amt const& req, Number const& dist)
 }
 // clang-format on
 
-/** Finds takerPays (i) and takerGets (o) such that given pool composition
- * poolGets(I) and poolPays(O): (O - o) / (I + i) = quality.
- * Where takerGets is calculated as the swapAssetIn (see below).
- * The above equation produces the quadratic equation:
- * i^2*(1-fee) + i*I*(2-fee) + I^2 - I*O/quality,
- * which is solved for i, and o is found with swapAssetIn().
+/** Solve quadratic equation to find takerGets or takerPays. Round
+ * to minimize the amount in order to maximize the quality.
+ */
+std::optional<Number>
+solveQuadraticEqSmallest(Number const& a, Number const& b, Number const& c);
+
+/** Generate AMM offer starting with takerGets when AMM pool
+ * from the payment perspective is IOU(in)/XRP(out)
+ * Equations:
+ * Spot Price Quality after the offer is consumed:
+ *     Qsp = (O - o) / (I + i) (1)
+ *  where O is poolPays, I is poolGets, o is takerGets, i is takerPays
+ * Swap out:
+ *     i = (I * o) / (O - o) * f (2)
+ *  where f is (1 - tfee/100000), tfee is in basis points
+ * Effective price targetQuality:
+ *     Qep = o / i (3)
+ * There are two scenarios to consider
+ * A) Qsp = Qep. Substitute i in (1) with (2) and solve for o
+ *    and Qsp = targetQuality(Qt):
+ *     o**2 + o * (I * Qt * (1 - 1 / f) - 2 * O) + O**2 - Qt * I * O = 0
+ * B) Qep = Qsp. Substitute i in (3) with (2) and solve for o
+ *    and Qep = targetQuality(Qt):
+ *     o = O - I * Qt / f
+ * Since the scenario is not know a priori, both A and B are solved and
+ * the lowest value of o is takerGets. takerPays is calculated with
+ * swap out eq (2). If o is less or equal to 0 then the offer can't
+ * be generated.
+ */
+template <typename TIn, typename TOut>
+std::optional<TAmounts<TIn, TOut>>
+getAMMOfferStartWithTakerGets(
+    TAmounts<TIn, TOut> const& pool,
+    Quality const& targetQuality,
+    std::uint16_t const& tfee)
+{
+    assert(targetQuality.rate() != beast::zero);
+    // calculate a, b, c to minimize quadratic equation solution
+    // b is negative
+    // calculate:
+    // (-b - root2(b * b - 4 * a * c)) / (2 * a)
+    // minimize:
+    // (-b - root2(b * b  - 4 * a * c)) / (2 * a)
+    // b
+    // maximize:
+    // root2(b * b - 4 * a * c)
+    // b * b
+    // minimize:
+    // 4 * a * c
+    // a
+    // c
+    // maximize:
+    // 2 * a
+    // a
+    // a must be maximized and minimized. a is 1, no rounding
+    // b must be maximized and minimized. choose minimize since the effect of
+    // the rounding outside of root2 is greater than the inside of root2.
```

Comment:
Except b is squared inside the root2. What you're saying may still be true, but it's not immediately clear to me that it is.

Reply:
To put it differently, we maximize the left side of the expression in several ways: `b * b` is maximized, `4 * a * c` is minimized, `b * b - 4 * a * c` is maximized and then the root is maximized as well. So I think the overall minimization effect is going to be greater if `b` is rounded downward. If this the right way, then `b` has to be actually rounded upward because `b` is negative in this case; i.e. we need to make the absolute `b` to be smaller.

Reply:
Here's what I did in sage:

```
sage: def q(a,b,c): return -b+sqrt(b*b-4*a*c)
sage: small_delta = 10^-20
sage: large_delta = 9*10^-20
sage: R=RealField(400)
sage: R(q(1,10+large_delta,1)) > R(q(1,10-small_delta,1))
True
```

I _think_ that shows that if we round up we can end up with larger values than if we round down. In this case, it's simulating rounding 10.00...001 up to 10.00...01 or down to 10.00...000). When I round `b` down I get a smaller value. It's possible I messed up this calculation (please dummy check me), but I'm still not convinced we'll find a single rounding for b that minimizes this. We may need one rounding inside the root2 and one outside the root2.

Reply:
A few things:
- The result is negative given `a,b,c` values.
- There is a relation between `b` and `c`. I think a real test should take it into consideration.
- How does this test factor in other rounding that I mentioned?
- Doesn't having different `b` kind of breaks the basic algebra? `b` must be the same in quadratic equation solution . x = (-b1 +- sqrt(b2*b2 - 4 * a *c)) / (2 * a) is not a valid quadratic equation solution even if `b1` and `b2` differ by a small delta. Am I missing something?

Reply:
I'm actually thinking how much of the rounding pays off in this function. The offer generation definitely must start with XRP and rounded downward. This makes a big difference, especially for small offers. But in case of IOU/IOU offer the rounding might not do much. To your earlier point, swap functions round in favor of the pool. IOU/IOU offer generation starts with takerPays and rounds it downward but takerGets is also rounded downward by swapAssetOut. As the result the AMM offer's quality might be less than the target quality. Which means we might have to allow for some error anyways and the rounding doesn't add much in this case except for more complexity.

Reply:
Just a general comment, not necessarily related to this bit of code:  I'm beginning to question the wisdom of directed rounding on every single operation.  The one other place I've seen this done is when one is trying to put lower and upper limits on the exact answer.  One does the computation twice, with consistent rounding one way, and then the other.

That's not what we're trying to do here.  What we want is to compute as accurately as possible, but make sure that round off error doesn't violate an invariant, such as creating XRP.  Perhaps we should stick to round_to_nearest, but then clip the final result to meet our invariants.  Perhaps clipping the final result, as opposed to rounding individual operations is the better tool here.

Reply:
This makes sense to me.

Reply:
I'm really poking at this statement:
```
b must be maximized and minimized. choose minimize since the effect of
the rounding outside of root2 is greater than the inside of root2
```
 It's not obvious to me that that's true, and I _think_ the test I wrote shows that it's not always true. If we really do want to choose the rounding that minimizes the whole expression sometimes we'll have to round up, and sometimes we'll have to round down (I _think_). There may be other constraints that I'm not considering (although I don't _think_ a relationship between b and c would matter, but maybe it does.)

So my real concern is that we always honor the invariant. If we can show that the invariant is honored in all cases, then we're good. An easy way to do that is to show we always set roundings to favor the AMM. If we can't do that then we have a much harder analysis job to do. (or we can give up trying to honor the invariant, but I'd rather not do that). Anyway, let's talk some more tomorrow about this.

Reply:
The invariant is honored in all cases. The reason for the rounding in this function is to minimize the generated offer so that the quality matches the provided target quality. This is accomplished by rounding downward the side of the offer that we start generating first. It doesn't matter which side it is - the smaller the amount the better the quality. The other side of the offer is calculated with either `swapAssetIn` or `swapAssetOut`. Those functions ensure the invariant is honored and round the result downward (to minimize takerGets) or upward (to maximize takerPays) respectively.

Reply:
As long as the invariant is honored I'm good. And I agree we want the most accurate result while honoring the invariant.

Reply:
I'm going to change the rounding per @HowardHinnant suggestion and address all current comments.